feat(core): implement a custom logging provider

ostridm opened the pull request:
closes #3

ostridm committed:
feat(core): implement a custom logging provider

closes #3

derevnjuk commented on src/SecTester.Core/Extensions/LoggerExtensions.cs:
```diff
@@ -0,0 +1,12 @@
+  public static void Error(this ILogger logger, Exception e)
+  {
+    logger.Error(e.ToString());
+  }
```
What is the point of having an extension when you can easily overload the method?

```csharp
  public void Error(Exception exception)
  {
    this.Error(exception.ToString());
  }
```

derevnjuk commented on src/SecTester.Core/Extensions/ServiceProviderExtensions.cs:
```diff
@@ -0,0 +1,12 @@
+
+public static class ServiceProviderExtensions
+{
+  public static T ResolveWith<T>(this IServiceProvider provider, params object[] parameters) where T : class
```
You don't use this helper. Please remove any unused classes. Anyway, it is better to extend `ServiceCollectionExtensions` by calling `AddLogging()` with the expected `ILoggerProvider`

derevnjuk commented on src/SecTester.Core/Logger/ColoredConsoleLogAppender.cs:
```diff
@@ -0,0 +1,52 @@
+
+      stream.WriteLine(message);
+
+      Console.ResetColor();
```
Why do you color a whole message instead of a header only? Why don't you use ANSI-coded colors with `ConsoleFormatter`?

derevnjuk commented on src/SecTester.Core/Logger/ColoredConsoleLogAppender.cs:
```diff
@@ -0,0 +1,52 @@
+
+public class ColoredConsoleLogAppender : ILogAppender
+{
+  private static SpinLock _consoleLock = new(false);
```
The spinlock will work exactly to the moment until you pass a huge message, which will lead to CPU spikes and delays.

derevnjuk commented on src/SecTester.Core/Logger/ILogger.cs:
```diff
@@ -0,0 +1,10 @@
+namespace SecTester.Core.Logger;
+
+public interface ILogger
```
Why don't use build-in abstractions instead? You just need to implement `ILoggerProvider`, `ILogger`, and `IConsoleFormatter`

derevnjuk replied:
The `I` prefix is forbidden (the same way `C` for classes, or `E` for enums).

derevnjuk commented on src/SecTester.Core/Logger/ILogAppender.cs:
```diff
@@ -0,0 +1,6 @@
+namespace SecTester.Core.Logger;
+
+public interface ILogAppender
```
ditto

derevnjuk commented on src/SecTester.Core/Utils/ISystemTimeProvider.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace SecTester.Core.Utils;
+
+public interface ISystemTimeProvider
```
ditto

derevnjuk commented on test/SecTester.Core.Tests/Logger/ColoredConsoleLogAppenderTests.cs:
```diff
@@ -0,0 +1,71 @@
+public class ColoredConsoleLogAppenderTests
+{
+  [Theory]
+  [InlineData("")]
+  [InlineData("message")]
```
Use `MemberData` attribute instead

derevnjuk commented on test/SecTester.Core.Tests/Logger/ColoredConsoleLogAppenderTests.cs:
```diff
@@ -0,0 +1,71 @@
+
+namespace SecTester.Core.Tests.Logger;
+
+[Collection("Sequential")]
```
You don't have collection definition class (i.e. `CollectionDefinition`)

derevnjuk commented on test/SecTester.Core.Tests/Logger/DefaultLoggerTests.cs:
```diff
@@ -0,0 +1,230 @@
+  private const string Message = "message";
+  private const string DatetimeFormat = "yyyy-MM-dd HH:mm:ss.fff";
+
+  private static readonly DateTime Moment;
+  private static readonly string HeaderLogLevelError;
+  private static readonly string HeaderLogLevelNotice;
+  private static readonly string HeaderLogLevelWarn;
+  private static readonly string HeaderLogLevelVerbose;
```
Why don't use private read-only instance members instead of class members?

derevnjuk commented on test/SecTester.Core.Tests/Logger/DefaultLoggerTests.cs:
```diff
@@ -0,0 +1,230 @@
+  private readonly ILogAppender _logAppenderMock;
+  private ServiceProvider _serviceProvider;
+
+  static DefaultLoggerTests()
```
ditto

derevnjuk commented on test/SecTester.Core.Tests/Logger/DefaultLoggerTests.cs:
```diff
@@ -0,0 +1,230 @@
+
+    // assert
+    _logAppenderMock.Received(1)
+      .WriteLine(LogLevel.Verbose, $"{DefaultLoggerTests.HeaderLogLevelVerbose}{DefaultLoggerTests.Message}");
```
`Message` is an instance member
```suggestion
      .WriteLine(LogLevel.Verbose, $"{DefaultLoggerTests.HeaderLogLevelVerbose}{Message}");

```

ostridm committed:
feat(core): implement a custom logging provider

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018290090
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018350905
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018736953
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018739498
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018743543
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018745043
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018745094
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018786771
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018789820
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018802658
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018802729
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018806790
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1018799084

closes #3

derevnjuk commented on src/SecTester.Core/Configuration.cs:
```diff
@@ -28,8 +32,10 @@ public class Configuration
+    public Configuration(string? hostname, Credentials? credentials = null, List<CredentialProvider>? credentialProviders = null, LogLevel logLevel = LogLevel.Error)
     {
+      LogLevel = logLevel;
+      
```
```suggestion
```

derevnjuk commented on src/SecTester.Core/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -15,5 +19,28 @@ public static IServiceCollection AddSecTesterConfig(this IServiceCollection coll
+    public static IServiceCollection AddDefaultLogging(this IServiceCollection collection, LogLevel logLevel = LogLevel.Error)
+    {
+      
+      collection.AddLogging((builder) =>
```
Remove redundant brackets

derevnjuk commented on src/SecTester.Core/Logger/DefaultConsoleFormatter.cs:
```diff
@@ -0,0 +1,109 @@
+
+
+  private readonly IDisposable _optionsReloadToken;
+  private DefaultConsoleFormatterOptions _formatterOptions;
```
I don't see any reason to couple them

```suggestion
  private ConsoleFormatterOptions _formatterOptions;

```

derevnjuk commented on src/SecTester.Core/Logger/DefaultConsoleFormatter.cs:
```diff
@@ -0,0 +1,109 @@
+  private readonly IDisposable _optionsReloadToken;
+  private DefaultConsoleFormatterOptions _formatterOptions;
+
+  protected SystemTimeProvider _systemTimeProvider;
```
It might be a private member

```suggestion
  private readonly SystemTimeProvider _systemTimeProvider;

```

derevnjuk commented on src/SecTester.Core/Logger/DefaultConsoleFormatter.cs:
```diff
@@ -0,0 +1,109 @@
+
+  public void Dispose()
+  {
+    _optionsReloadToken?.Dispose();
```
You can remove redundant conditional access operator
```suggestion
    _optionsReloadToken.Dispose();

```

derevnjuk commented on src/SecTester.Core/Logger/DefaultConsoleFormatter.cs:
```diff
@@ -0,0 +1,109 @@
+
+  protected virtual void WriteHeader<TState>(
+    in LogEntry<TState> logEntry,
+    IExternalScopeProvider? scopeProvider,
```
Unused parameter
```suggestion

```

derevnjuk commented on src/SecTester.Core/Logger/DefaultConsoleFormatter.cs:
```diff
@@ -0,0 +1,109 @@
+
+  public override void Write<TState>(
+    in LogEntry<TState> logEntry,
+    IExternalScopeProvider? scopeProvider,
```
`scopeProvider` is not nullable according to signature

```suggestion
    IExternalScopeProvider scopeProvider,

```

derevnjuk commented on src/SecTester.Core/Logger/ColoredConsoleFormatter.cs:
```diff
@@ -0,0 +1,46 @@
+  static string GetForegroundColorAnsiCode(LogLevel level) =>
+    level switch
+    {
+      LogLevel.Critical  => "\x1B[1m\x1B[31m",  //ConsoleColor.Red
```
The same comment relates to each statement below
```suggestion
      LogLevel.Critical  => "\x1B[1m\x1B[31m",  // ConsoleColor.Red

```

derevnjuk commented on src/SecTester.Core/Logger/DefaultConsoleFormatterOptions.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace SecTester.Core.Logger;
+
+public class DefaultConsoleFormatterOptions: ConsoleFormatterOptions
```
Redundant class, you can remove it without any issues. Use `ConsoleFormatterOptions` instead

derevnjuk commented on src/SecTester.Core/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -15,5 +19,28 @@ public static IServiceCollection AddSecTesterConfig(this IServiceCollection coll
+      
+      collection.AddLogging((builder) =>
+      {
+        builder.AddConsoleFormatter<ColoredConsoleFormatter, DefaultConsoleFormatterOptions>();
```
For details see the comments above


```suggestion
        builder.AddConsoleFormatter<ColoredConsoleFormatter, ConsoleFormatterOptions>(options => {
          options.TimestampFormat = "yyyy-MM-dd HH:mm:ss.fff";
          // etc
        });
```

derevnjuk commented on src/SecTester.Core/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -15,5 +19,28 @@ public static IServiceCollection AddSecTesterConfig(this IServiceCollection coll
       return collection;
     }
+
+    public static IServiceCollection AddSystemTimeProvider(this IServiceCollection collection, SystemTimeProvider? instance = null)
```
Should we provide such atomic control of settings? I propose to register all of them with default options in `AddSecTesterConfig`, e.g.

```csharp
public static IServiceCollection AddSecTesterConfig(this IServiceCollection collection, string hostname)
{
  AddSecTesterConfig(collection, new Configuration(hostname));
  return collection;
}

public static IServiceCollection AddSecTesterConfig(this IServiceCollection collection, Configuration configuration)
{
  collection.Add(new ServiceDescriptor(typeof(Configuration), configuration));
  collection.AddSingleton(new UtcSystemTimeProvider());
  collection.AddLogging((builder) =>
  {
    // ...
    builder.SetMinimumLevel(configuration.LogLevel);
    // ...
  });
  return collection;
}
```

derevnjuk commented on src/SecTester.Core/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -15,5 +19,28 @@ public static IServiceCollection AddSecTesterConfig(this IServiceCollection coll
+      collection.AddLogging((builder) =>
+      {
+        builder.AddConsoleFormatter<ColoredConsoleFormatter, DefaultConsoleFormatterOptions>();
+        builder.SetMinimumLevel(logLevel);
+        builder.AddConsole(options =>
```
Please use a chain of calls where it is possible

```suggestion
        builder.SetMinimumLevel(logLevel).AddConsole(options =>
```

derevnjuk commented on test/SecTester.Core.Tests/Logger/ColoredConsoleFormatterTests.cs:
```diff
@@ -0,0 +1,51 @@
+    new object[] { LogLevel.Debug, "\x1B[1m\x1B[37m" }, new object[] { LogLevel.Trace, "\x1B[1m\x1B[36m" }
+  };
+
+  static LogEntry<Tuple<string, object[]>> CreateLogEntry(LogLevel logLevel, string message, params object[] args)
```
```suggestion
  private static LogEntry<Tuple<string, object[]>> CreateLogEntry(LogLevel logLevel, string message, params object[] args)

```

derevnjuk commented on test/SecTester.Core.Tests/Logger/ColoredConsoleFormatterTests.cs:
```diff
@@ -0,0 +1,51 @@
+  {
+    return new LogEntry<Tuple<string, object[]>>(logLevel, "category", 1, new Tuple<string, object[]>(message, args),
+      null,
+      (state, exception) => exception != null ? exception.ToString() : string.Format(state.Item1, state.Item2));
```
```suggestion
      (state, exception) => exception?.ToString() ?? string.Format(state.Item1, state.Item2));

```

derevnjuk commented on test/SecTester.Core.Tests/Logger/DefaultConsoleFormatterTests.cs:
```diff
@@ -0,0 +1,102 @@
+    },
+  };
+
+  static LogEntry<Tuple<string, object[]>> CreateLogEntry(LogLevel logLevel, string message, params object[] args)
```
```suggestion
  private static LogEntry<Tuple<string, object[]>> CreateLogEntry(LogLevel logLevel, string message, params object[] args)

```

derevnjuk commented on test/SecTester.Core.Tests/Logger/DefaultConsoleFormatterTests.cs:
```diff
@@ -0,0 +1,102 @@
+  {
+    return new LogEntry<Tuple<string, object[]>>(logLevel, "category", 1, new Tuple<string, object[]>(message, args),
+      null,
+      (state, exception) => exception != null ? exception.ToString() : string.Format(state.Item1, state.Item2));
```
```suggestion
      (state, exception) => exception?.ToString() ?? string.Format(state.Item1, state.Item2));

```

ostridm committed:
feat(core): implement a custom logging provider

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021157110
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021154018
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021153240
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021147122
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021147990
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021148991
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021150956
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021151814
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021163707
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021164149
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021163232
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021162849
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021159927
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021159470
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021151814
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021142472
https://github.com/NeuraLegion/sectester-net/pull/33#discussion_r1021143129

closes #3
